### DISCUSSION TITLE
Language file - Tamil, an Indian language

### DISCUSSION THREAD

**Discussion by pvanthony** [romcal/romcal#378] [General]
Hi,

Friar Julian has done some more of the translation but there are still more to go.

Could we upload the file so that others can check and contribute with the remaining words.
[ta.zip](https://github.com/romcal/romcal/files/9694368/ta.zip)

P.V.Anthony

**Comment by tukusejssirs** [comment]
@pvanthony, while we can do it instead of you, would you like to learn how create a pull request (PR) and all stuff needed to merge it into romcal? IMHO a work-in-progress/draft PR is the best to track changes of a feature.

Moreover, we need to update it based on PR #350 which will be merged after @emagnier finds some time to review it, as I fixed some localisation issues  in there. Again I could that myself if you don’t want to create a PR yourself.

Also, I think that untranslated keys should be commented out, as they would be used from `en.ts` nonetheless and (for now) they would needlessly increase the size of the romcal source, as well as build size.

Otherwise nice work! Thank you very much! :pray:

**Reply by tukusejssirs** [reply]
@pvanthony, I created PR #379. I invite you (and potentially Friar Julian) to continue the discussion over there.

Thank you once again for the great work! :pray:

**Comment by pvanthony** [comment]
Thank you for doing the "PR". I am still learning git. Long way to go.
We are still updating the translation. Will update once we have done more.
Once again thank you for the patience.

**Reply by tukusejssirs** [reply]
@pvanthony, if you need help with anything related to Git, feel free to ping me and I’ll help you. :wink:

**Reply by pvanthony** [reply]
Sorry for the delay. Only just got some time to look into romcal.

Need some advice how to use the latest version of romcal.
Currently we have https://missal.org. It has been running for sometime now.
How to find current version on https://missal.org?

Can we just use the following in our index.html page?
```<script src="https://unpkg.com/romcal@3.0.0-dev.29/dist/iife/romcal.js"></script>```
Is this the latest version? If not, how to find the latest version? Or should I need to "compile" it?

Once the above works for English, we would like use the Tamil_Indian translation. Need advice on how to do that.
Like which file to call in the index.html page under script.

Lastly need advice on how to use our own country settings. Like Singapore. We have not submitted it but want to try and use it.
I am trying not to 'compile' and use the cdn version. Or because we are using our own unsubmitted country calendar, do we have to 'compile' romcol?

I am asking all above because previously, I would download from github and 'compile' so the calendar changes and language was all done in the 'compile'. Please note I do not know if 'compile' is the correct word to use.

I prefer not to 'compile' and just use what is available from unpkg.com. I am being lazy. :-) 

I am hoping it will not be a rewrite for me at https://missal.org :-)

Thank you for bearing with a newbie.

**Reply by TobiTenno** [reply]
Yes, you can leverage that script style for your website.

The [Releases](https://github.com/romcal/romcal/tags) page will have the latest version, you'll want the latest prerelease whenever you have time to update the version.

For others' reference, the instructions for installing or importing romcal are [here](https://github.com/romcal/romcal#1-import-romcal)

**Reply by pvanthony** [reply]
Thank you very much for the advice. 
Doing some testing with the following code on a html file but I getting an error in the console.

This is the script in the html file.
```
<script src="https://unpkg.com/romcal@3.0.0-dev.79/dist/iife/romcal.js"></script>

<script>
const romcal = new Romcal();
const dates = romcal.dates(2030);
const easterOf2030 = dates.easterSunday();
</script>

```
But getting the following error. 
```
Uncaught TypeError: Romcal is not a constructor
```
Seems to caused by "const romcal = new Romcal();"
What have I missed? Really need advice.

Using firefox 128.0.3 on fedora 40 linux.

**Reply by TobiTenno** [reply]
try `new Romcal.Romcal()`, i believe we have that was a change on about dev.39

It worked for me when i tested in an html file

**Reply by pvanthony** [reply]
> try `new Romcal.Romcal()`, i believe we have that was a change on about dev.39
> 
> It worked for me when i tested in an html file

Yes! It works for me too.
Thank you for the solution.

**Reply by pvanthony** [reply]
Got the romcal data. Now for the country specific calendar.
We have one for Singapore. Please note, we are not ready to submit it because it may not be correct for now.
Anyway here is the contents of singapore.ts

```ts
import dayjs from 'dayjs';

import * as Locales from '@romcal/lib/locales';
import { LiturgicalColors } from '@romcal/constants/liturgical-colors/liturgical-colors.enum';
import { LiturgicalDayInput } from '@romcal/models/liturgical-day/liturgical-day.types';
import { RomcalConfig, RomcalConfigInCalendarDef } from '@romcal/models/config/config.model';
import { Ranks } from '@romcal/constants/ranks/ranks.enum';
import { Titles } from '@romcal/constants/titles/titles.enum';

const defaultConfig: RomcalConfigInCalendarDef | undefined = undefined;

const dates = async (config: RomcalConfig): Promise<Array<LiturgicalDayInput>> => {
  const year = config.year;
  const _dates: Array<LiturgicalDayInput> = [
    {
      key: 'martha_of_bethany_mary_of_bethany_and_lazarus_of_bethany',
      extend: true,
      prioritized: true,
    },
    {
      key: 'our_lady_of_sorrows',
      extend: true,
      prioritized: true,
    },
    {
      key: 'therese_of_the_child_jesus_and_the_holy_face_of_lisieux_virgin',
      extend: true,
      prioritized: true,
    },
    {
      key: 'guardian_angels',
      extend: true,
      prioritized: true,
    },
    {
      key: 'francis_xavier_priest',
      extend: true,
      prioritized: true,
    },
    {
      key: 'barnabas_apostle',
      extend: true,
      prioritized: true,
    },
    {
      key: 'passion_of_saint_john_the_baptist',
      extend: true,
      prioritized: true,
    },
  ];

  // Get localized liturgical day names
  return await Locales.localizeDates(_dates);
};

export { dates, defaultConfig };
```

I am using the following in the html so that I do not have to compile romcal.
```
<script src="https://unpkg.com/romcal@3.0.0-dev.79/dist/iife/romcal.js"></script>
```

How do I compile singapore.ts alone? Or do have to compile the whole romcal?

Later I would also need to compile Tamil language too.

**Reply by tukusejssirs** [reply]
@pvanthony, I think it might be better for you to compile romcal until you notify us your changes (changes to the Tamil locale and the Singapore calendar) are ready to get merged into romcal. While compiling a single file might be possible (albeit not trivial), I see no advantage doing so in comparison with full compilation of romcal. I think it would be much easier for you to compile romcal with `singapore.ts` added and [`lib/particular-calendars/index.ts`](https://github.com/romcal/romcal/blob/dev/lib/particular-calendars/index.ts) updated. (Maybe @emagnier has different thoughts on this.)

Regarding the Singapore calendar, I have found a few, rather small issues:

- some imports are unused (e.g. `dayjs`);
- `extend` property is unnecessary any more;
- there are some changes in how the calendar is defined;
- I have no idea if some celebrations are celebrated throughout Asia (e.g. in Europe or the Americas, there are continental patrons which are celebrated in the whole continent); if there are some, those celebrations should be moved to a separate `asia.ts` calendar and `singapore.ts` should have `ParentCalendar` set to `Asia` (see [`lib/particular-calendars/england.ts`](https://github.com/romcal/romcal/blob/dev/lib/particular-calendars/england.ts)):
   - if there is no `ParentCalendar` defined, the General Roman Calendar will be used as the parent calendar;
- no celebration should be prioritised by default;
- if a celebration is already defined in the General Roman Calendar or one of the parent calendars (e.g. in `asia.ts`), the celebration should either not be included in `singapore.ts` or only changed properties should be listed:
   - all the celebration you have shared with us in the calendar are already defined in [`lib/general-calendar/proper-of-saints.ts`](https://github.com/romcal/romcal/blob/dev/lib/general-calendar/proper-of-saints.ts) and you have not added any changed properties to the celebrations (apart for the unnecessary `extend` and `prioritized` properties), no need to define any of these celebrations in `singapore.ts`;
- `guardian_angels` were renamed to `holy_guardian_angels`:
   - for all currently used celebration IDs, see [`lib/locales/en.ts`](https://github.com/romcal/romcal/blob/dev/lib/locales/en.ts).

Building romcal is very easy, although I get it that using a CDN is easier if you use a shared hosting for your website with no command-line access to the server. Still, you can build romcal locally on your computer and copy the `dist/iife/romcal.js` file to your server.

```bash
# Clone or download the latest `dev` romcal version into `romcal` folder in the current folder
# Note: Run only one of the following commands.
git clone https://github.com/romcal/romcal.git -b v3.0.0-dev.79
mkdir romcal && curl -L https://github.com/romcal/romcal/archive/refs/tags/v3.0.0-dev.79.tar.gz | tar xz -C romcal --strip-components 1

# Install romcal dependencies
npm i

# [optional] Do some changes to the codebase

# Build romcal
# Note: Run this command after any and all changes to the codebase
npm run build

# Copy this file to you website host and replace the unpkg.com links with a path to this file on the website host
dist/iife/romcal.js
```

While I have already asked you for this [here](https://github.com/romcal/romcal/pull/379#pullrequestreview-1820294851), but I wish you would provide us the sources for all Singaporean celebrations and Tamil localisation (ideally a text version, but images/photos will do too). While I don’t speak Tamil, I think I am good at analysing languages. The most easily accessible source is the Tamil translation of the Roman Missal where there is a list of all celebrations inscribed in the national calendar (including those inscribed in the GRC which would need to be removed when adding that to romcal). Some celebration names will need to be sourced from other pages of the Roman Missal (e.g. `commemoration_of_all_the_faithful_departed`), however, that could be done in the next step. @pvanthony, would you be able to create photos of some ~15 pages from the Tamil translation of the Roman Missal? You could ask your parish priest for permission. You could also let us contact the priest (ideally via email or Google Meet / Zoom). The best form would be a searchable PDF file, however, I am no sure if someone has created one for the Tamil translation of the Roman Missal and there might be copyright issues (we would still welcome it as an internal resource shared only with the core romcal developers).

The reason why I keep asking for the sources is that I think we should add source references to each piece of data (celebration or localisation), thus to provide verified and verifiable data. Without the access to the sources it is hard to achieve, however, some sources are more easily accessed by a local person, and the Roman Missal is one of them. I personally consider this also as a way to speed up the addition of the Tamil locale and the Singaporean calendar to romcal.

> I am hoping it will not be a rewrite for me at https://missal.org/ :-)

I am not sure about this one. While there are for sure some l10n key changes, other than that I think your website will not need to be rewritten. That said, I suggest to add the celebration names to each event/day instead of just providing a date and a weekday name. YMMV. :wink:

**Reply by pvanthony** [reply]
Thank you for the reply and the instructions on how to compile.
The instructions worked well.
What I did was to take from your repo the ta.js file and placed it in the 79 version. Then compiled it and it works.
Still need advice on how to get the translation for the following code but only the name key comes.
```
isoTodayDate = today.getFullYear() + "-" + String(today.getMonth()+1).padStart(2, '0') + "-" + String(today.getDate()).padStart(2, '0');
todayArray = result[isoTodayDate];
  console.log(todayArray[0].name) <-- this works and is translated
  console.log(todayArray[0].cycles.weekdayCycle) <-- works but not translated
  console.log(todayArray[0].rank) <-- works but not translated
```
Is there some other key I have to call? I have tested with French and still the same result as Tamil. Also used the compile from unpkg. Same behaviour.

Once this is working, I will do the corrections for the Singapore calendar to work.

Regarding the documentation for the translation, let me ask Friar Julian. I am not sure if he has any. He serves mass in Tamil so I wonder if he just got them by heart. Anyway I will discuss with him. Hopefully he gives permission for his gmail email to be sent to you.

**Reply by tukusejssirs** [reply]
> What I did was to take from your repo the ta.js file and placed it in the 79 version. Then compiled it and it works.

I have just rebased my [`tamil`](https://github.com/tukusejssirs/romcal/tree/tamil) branch on to of `dev` branch, thus all you should need to do is that branch to build romcal again. No need to copy `ta.ts` anywere, just download/clone my `tamil` branch and build it.

> Still need advice on how to get the translation for the following code but only the name key comes.

You need to append `Name` to the property names:

```js
todayArray[0].colorNames
todayArray[0].cycles.properCycleName
todayArray[0].cycles.psalterWeekName
todayArray[0].cycles.sundayCycleName
todayArray[0].cycles.weekdayCycleName
todayArray[0].name
todayArray[0].rankName
todayArray[0].seasonNames
```

> Regarding the documentation for the translation, let me ask Friar Julian. I am not sure if he has any. He serves mass in Tamil so I wonder if he just got them by heart. Anyway I will discuss with him. Hopefully he gives permission for his gmail email to be sent to you.

If Fr Julian says the Holy Mass in Tamil, he definitely has a Tamil translation of the Roman Missal. :wink: I have just found [this website](https://bibleintamil.com/missal/index.html) where there is an image of the Tamil Missal. :wink:

We will be very grateful if you could provide us with the sources. :pray:

Anyway Fr Julian or you can send the PDFs/images to romcal@romcal.dev. Then, the romcal core developers will have access to it.

**Reply by pvanthony** [reply]
> I have just rebased my [`tamil`](https://github.com/tukusejssirs/romcal/tree/tamil) branch on to of `dev` branch, thus all you should need to do is that branch to build romcal again. No need to copy `ta.ts` anywere, just download/clone my `tamil` branch and build it.

Thank you for this. I will use your repo from now onwards

> You need to append `Name` to the property names:

This works great. Looking great.

I will start working on the Singapore calendar.




> If Fr Julian says the Holy Mass in Tamil, he definitely has a Tamil translation of the Roman Missal. 😉 I have just found [this website](https://bibleintamil.com/missal/index.html) where there is an image of the Tamil Missal. 😉
> 
> We will be very grateful if you could provide us with the sources. 🙏
> 
> Anyway Fr Julian or you can send the PDFs/images to [romcal@romcal.dev](mailto:romcal@romcal.dev). Then, the romcal core developers will have access to it.

I will be meeting up with him next week. Will definitively inform him and update you about the sources.

**Reply by emagnier** [reply]
> I think it would be much easier for you to compile romcal with singapore.ts added and [lib/particular-calendars/index.ts](https://github.com/romcal/romcal/blob/dev/lib/particular-calendars/index.ts) updated. (Maybe @emagnier has different thoughts on this.)

I confirm, it’s currently not possible to properly use external data sources with romcal. For now, the cleanest method remains to install and compile the romcal project directly, with singapore.ts added.
